### PR TITLE
Partial processing

### PR DESCRIPTION
**Pull request by adon-at-work**
- added partial processing, i.e., parse the partial content for contextual analysis
- ~~configurable to completely skip any partial processing using enablePartialProcessing = false~~ 
- running secure-handlebars without providing partial directory (i.e., by default) means to skip partial processing
- ~~allowed possibility to individually skip missing partial check using {{> SJST/SKIP/...}}~~
- corrected regular expression to match partialName to allow the slash (/) char
- resolved partialName using path.relative instead of regular expression
- allowed partial lookup in deeper folder, as provided by partialsDir
- generated partials are all located in a folder called SJST

### CODE REVIEW

File: bin/handlebarspp [hunk 1]
```diff
@@ -13,29 +13,120 @@ This utility parses the handlebars template file and adds the context-sensitive
...
-                strictMode: process.argv[3] === 'true',
-                processingFile: file
+    function printUsage() {
+        console.log("Usage: handlebarspp template...\n\n\
+Options:\n\
+\t-D\tDisable Partial Handling, and assume that partials are placed and will result in the text (aka HTML Data) context (default: false).\n\n\
+\t-p\tPartial Directory. The Secure Handlebars preloads the partials into the memory for partial parsing.\n\n\
+\t-e\tPartial File Extension. The partial file extension. (default: .handlebars)\n\n\
+\t-c\tMelt Partials into Template (experimental). The Secure Handlebars combines the partial contents into the input template. (default:false)\n\n\
+\t-d\tDump the preprocessed partials (prefixed with SJST/) into the partial directory\n\n\
```
Comment (neraliu) [inline bin/handlebarspp]: if u dump the combined template into this default folder SJST, please make it configurable.

File: src/handlebars-utils.js [hunk 2]
```diff
@@ -72,8 +72,9 @@ HandlebarsUtils.ESCAPE_EXPRESSION = 2; // {{expression}}
 HandlebarsUtils.escapeExpressionRegExp = /^\{\{~?\s*@?\s*([^\}\{~]+)~?\}\}(?!})/;
 
 /* '{{' '~'? '>' '\s'* ('not \s, special-char'+) '\s'* 'not ~{}'* non-greedy '}}' and not follow by '}' */
+/* slash should be allowed */
 HandlebarsUtils.PARTIAL_EXPRESSION = 3; // {{>.*}}
-HandlebarsUtils.partialExpressionRegExp = /^\{\{~?>\s*([^\s!"#%&'\(\)\*\+,\.\/;<=>@\[\\\]\^`\{\|\}\~]+)\s*[^~\}\{]*~?\}\}(?!})/;
+HandlebarsUtils.partialExpressionRegExp = /^\{\{~?>\s*([^\s!"#%&'\(\)\*\+,\.;<=>@\[\\\]\^`\{\|\}\~]+)\s*([^~\}\{]*)~?\}\}(?!})/;
```
Comment (neraliu) [inline src/handlebars-utils.js]: please add a comment to the code of why removing `/` in partialExpressionRegExp.

**Comment by neraliu** [comment]
> allowed possibility to individually skip missing partial check using {{> SJST/SKIP/...}}

this should be configurable if supporting partial skipping

**Comment by neraliu** [comment]
> configurable to completely skip any partial processing using enablePartialProcessing = false
i prefer this feature to be disable by default first before it is stable. and if it is disable, SH should report the warning message right?

File: bin/handlebarspp (hunk 1 above)
Comment (adon-at-work) [inline bin/handlebarspp]: this is for the preprocessed partials only.
do you mean you want to enhance it by dumping the template with partial combined?

File: src/handlebars-utils.js (hunk 2 above)
Comment (adon-at-work) [inline src/handlebars-utils.js]: em... i guess a comment is not needed: i believe handlebars allows using `/` among other chars in partialExpression. i think it's actually a bug when `/` was disallowed.

here's an example in which `express-handlebars` is using `/`, and it functions properly with the vanilla handlebars.
https://github.com/ericf/express-handlebars/blob/master/examples/advanced/views/echo.handlebars

I guess it's more like a TODO, can you pls check if your regexp is over restrictive, or if it is intended?

**Comment by adon-at-work** [comment]
> > allowed possibility to individually skip missing partial check using {{> SJST/SKIP/...}}

> this should be configurable if supporting partial skipping

do you mean to configure the keyword `SKIP`?

**Comment by adon-at-work** [comment]
> > configurable to completely skip any partial processing using enablePartialProcessing = false

> i prefer this feature to be disable by default first before it is stable. and if it is disable, SH should report the warning message right?

we can turn it off by default. once turned off, no more warning... seems we need another config to suppress warning?

**Comment by adon-at-work** [comment]
@neraliu, this is ready for another round of reviews.

**Pull request merged**